modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/UnionConverterRule.java

```diff
@@ -48,16 +48,18 @@ public UnionConverterRule(Config cfg) {
     /** {@inheritDoc} */
     @Override
     public void onMatch(RelOptRuleCall call) {
-        final LogicalUnion union = call.rel(0);
+        LogicalUnion union = call.rel(0);
 
         RelOptCluster cluster = union.getCluster();
         RelTraitSet traits = cluster.traitSetOf(IgniteConvention.INSTANCE);
+
         List<RelNode> inputs = Commons.transform(union.getInputs(), input -> convert(input, traits));
+        List<RelNode> convertedInputs = Commons.castInputsToLeastRestrictiveTypeIfNeeded(inputs, cluster, traits);
```

Comment:
i change a bit this method implementation :
if (resultType.equalsSansFieldNames(inputRowType) || equalAsStructSansNullability(typeFactory(), resultType, inputRowType, null)) {
and seems all pruning tests are ok and intergation too, wdyt ?

Reply:
Yes. Conversion from not nullable to nullable is always safe.
This PR should not touch partition pruning.